Add tracking for failed jobs and custom callbacks for each step in lifecycle

sunrick opened the pull request:
# Failed Jobs

You can now track jobs that have failed!
# Custom callback methods

I've defined custom methods that are overridable. Previously, writing code in `before_perform` block in child class would override active job status `before_perform` block. Custom methods can be used without ruining tracking functionality. You can also handle scenarios where job code fails.

sunrick committed:
Add failed job support and overridable custom callback methods using active job callbacks.

sunrick committed:
Add basic documentation for new features.

brad-johnson-digisight commented on lib/active_job_status/hooks.rb:
```diff
@@ -2,12 +2,54 @@ module ActiveJobStatus
+          rescue Exception => exception
+            job_tracker.failed
+            on_enqueue_failure(exception, job)
+            raise
```
I'm a bit concerned that we might be swallowing useful information here. Perhaps I'm mistaken, but if a user does not supply a custom `on_enqueue_failure` block, all exceptions will be caught, and the actual exception will get swallowed, and line 15 will raise a basic exception instead.

cdale77 commented on lib/active_job_status/hooks.rb:
```diff
@@ -2,12 +2,54 @@ module ActiveJobStatus
+            job_tracker.enqueued
+            block.call
+            on_enqueue_success(job)
+          rescue Exception => exception
```
I'm a bit concerned that we might be swallowing useful information here. Perhaps I'm mistaken, but this seems to catch all exceptions, pass them off to a block (if supplied), and then re-raises a generic exception. At a minimum we'd need to re-raise the exception that was actually thrown, so this doesn't swallow them.

cdale77 commented:
@sunrick thanks!! There are some merge conflicts, It think mostly because I fixed https://github.com/cdale77/active_job_status/issues/17. 

I like the idea of providing information about failed jobs. 

I'm concerned about having ActiveJobStatus provide job lifecycle callback functionality, especially regarding catching exceptions. It just seems like not the right fit for this library, and ActiveJob already provides it. 

But it *does* seem pretty problematic that this library is overriding existing callbacks, thanks for pointing that out! It seems to me the fix is making ActiveJobStatus respect the existing callback functionality and not override those methods. I'm not sure off the top of my head how to do that. Thoughts?

boie0025 commented:
👍

jrochkind commented:
This is super useful functionality. I wonder if it makes sense to store the exception class and/or backtrace too?

I wonder if in general returning just a symbol job status is insufficient. Other libraries say something is `complete?` whether it's successful or failed, and have separate checks for succesful or failure.  It might also be interesting to store job start and finish times.

sunrick committed:
Fix typo in documentation.

sunrick committed:
Explicity reraise exception.

sunrick committed:
Merge branch 'master' of github.com:cdale77/active_job_status into failed-jobs

sunrick committed:
Remove pry from development.

sunrick committed:
Fix formatting.

sunrick commented:
@cdale77 I updated the branch to a better state than it was before. Fixed merge conflicts. I also explicitly re-raise the exception.  (it technically worked before) 

There is a test that doesn't pass right now. 

```ruby
     it "should be true when jobs are not in the store" do
       clear_store(id_array: first_jobs)
       expect(batch.completed?).to be_truthy
     end
```
With the way the fetch and get status methods are written they will return nil in two different scenarios: 1) the status is empty 2) the job doesn't exist in the store.
```ruby
module ActiveJobStatus
  class NoStoreError < StandardError; end

  class << self
    attr_accessor :expiration
    attr_writer :store

    def get_status(job_id)
      fetch(job_id).status
    end

    def fetch(job_id)
      status = store.fetch(job_id)
      JobStatus.new(status)
    end

    def store
      unless @store
        raise NoStoreError, "can't use ActiveJobStatus without store"
      end
      @store
    end
  end
end
```
Can't make the test pass because of that behavior, unless I'm missing something.

I don't actually use the library any more so if someone wants to add features/clean things up please feel free to do so!
